Changes to use glbinding 3.0.2 and Forge 1.0.3

9prady9 opened the pull request:
Fixes #2354 

- Forge 1.0.3 uses glbinding 3.0.2
- ArrayFire's glbinding has also been bumped up to 3.0.2

9prady9 commented:
build are going to fail until some changes are made in the ci systems.

umar456 commented on CMakeModules/build_forge.cmake:
```diff
@@ -36,7 +36,7 @@ ExternalProject_Add(
       -DBUILD_SHARED_LIBS:BOOL=ON
       -DCMAKE_INSTALL_PREFIX:PATH=<INSTALL_DIR>
-      -DCMAKE_BUILD_TYPE:STRING=Release
+      -DCMAKE_BUILD_TYPE:STRING=${CMAKE_BUILD_TYPE}
```
I think its its appropriate to keep this as release. We are using the library not debugging it.

umar456 commented on src/backend/common/CMakeLists.txt:
```diff
@@ -105,16 +105,25 @@ if(AF_WITH_GRAPHICS)
         $<TARGET_FILE:glbinding::glbinding>
         $<$<PLATFORM_ID:Linux>:$<TARGET_SONAME_FILE:glbinding::glbinding>>
         $<$<PLATFORM_ID:Darwin>:$<TARGET_SONAME_FILE:glbinding::glbinding>>
+        $<TARGET_FILE:glbinding::glbinding-aux>
```
These are only need for debugging right?

9prady9 commented:
aux library is not needed in release mode but then I forgot to enclose that in if block because I was thinking about our release binaries build type - RelWithDebInfo.

9prady9 commented on src/backend/common/CMakeLists.txt:
```diff
@@ -64,6 +64,7 @@ target_link_libraries(afcommon_interface
 
 target_include_directories(afcommon_interface
   INTERFACE
+    ${OPENGL_INCLUDE_DIR}
```
I added that include folder in wrong location. We need that inside the IF block based on `AF_WITH_GRAPHICS`.

9prady9 commented on CMakeLists.txt:
```diff
@@ -216,26 +245,23 @@ if(Forge_FOUND AND NOT AF_USE_SYSTEM_FORGE)
+    set(fg_bin_dir "/extern/forge/package/")
     if(AF_INSTALL_FORGE_DEV)
-        install(DIRECTORY "${ArrayFire_BINARY_DIR}/third_party/forge/include/"
+        install(DIRECTORY "${ArrayFire_BINARY_DIR}${fg_bin_dir}include/"
```
There is one issue with that though, `make install` on forge project pulls all dependencies and installs them in `package/lib` (assuming package is the install prefix). If we use current route, the graphics dependencies are also pulled automatically while building packages. If we use target based installation, then we can to write logic for pulling glfw3, glbinding again.

9prady9 commented on CMakeLists.txt:
```diff
@@ -92,7 +91,37 @@ mark_as_advanced(
 
 if(AF_WITH_GRAPHICS AND NOT AF_USE_SYSTEM_FORGE)
-  include(build_forge)
+    set(ArrayFireInstallPrefix ${CMAKE_INSTALL_PREFIX})
```
Sure

9prady9 commented on CMakeLists.txt:
```diff
@@ -92,7 +91,37 @@ mark_as_advanced(
+        glm_DIR
+        glbinding_DIR
+        )
+    add_custom_command(
```
`install` target doesn't run automatically in submodule targets from what I have experienced. For the reason mentioned in previous comment, I needed a way to run `install` script so that dependencies from forge in once place.

9prady9 commented:
I am second guessing the route of using forge as submodule. Lets discuss this once more today.

umar456 commented on CMakeLists.txt:
```diff
@@ -212,33 +206,7 @@ install(FILES ${ArrayFire_BINARY_DIR}/include/af/version.h
-      DESTINATION "${AF_INSTALL_LIB_DIR}"
-      COMPONENT common_backend_dependencies)
-endif()
+#TODO(pradeep) install forge dependency for packaging - not required for builds
```
stale TODO?

9prady9 replied:
nope, I am not sure how cpack stuff for graphics deps will be handled. So added this TODO to make sure we don't forget it.

umar456 replied:
I wonder if that sort of thing should be added to cpackconfig.cmake

umar456 commented on CMakeModules/AFconfigure_forge_submodule.cmake:
```diff
@@ -0,0 +1,28 @@
+set(FG_BUILD_EXAMPLES OFF CACHE BOOL "Used to build Forge examples")
+set(FG_BUILD_DOCS OFF CACHE BOOL "Used to build Forge documentation")
+set(FG_WITH_FREEIMAGE OFF CACHE BOOL "Turn on usage of freeimage dependency")
+add_subdirectory(extern/forge EXCLUDE_FROM_ALL)
```
`add_subdirectory` should stay in the top level CMake. Also it would be nice if this was in a function

9prady9 replied:
Is this any different ? because `include` basically just replaces the content into here right ?

umar456 replied:
Yes but it kind of hides the details. It would be better to have all the subdirectories in one location

umar456 commented on CMakeModules/platform.cmake:
```diff
@@ -24,10 +27,8 @@ if(UNIX AND NOT APPLE)
   # C4275: Warnings about using non-exported classes as base class of an
   #        exported class
-  add_compile_options(/wd4251 /wd4068 /wd4275)
+  add_compile_options(/wd4068 /wd4275)
```
:+1:

9prady9 commented:
@umar456 I have removed `AF_WITH_GRAPHICS` and added `AF_BUILD_FORGE`. You can review the most recent commit to see the details of what the new cmake option does. I have addressed all changes completely except the file `CMakeModules/osx_install/OSXInstaller.cmake` - I have made some changes that seemed alright to me, but I need your help with OSX specific things you did in that file.

9prady9 commented:
After some discussions on slack channel, the files under `CMakeModules/osx_install` are obsolete except couple of html files. These have been removed too in this PR.

umar456 commented on CMakeModules/AFconfigure_forge_submodule.cmake:
```diff
@@ -0,0 +1,40 @@
+    FG_ENABLE_HUNTER
+    glfw3_DIR
+    glm_DIR
+    glbinding_DIR
```
This is not required right?

umar456 commented on CMakeModules/CPackConfig.cmake:
```diff
@@ -317,7 +313,7 @@ set(CPACK_RPM_PACKAGE_AUTOREQPROV " no")
 set(CPACK_RPM_PACKAGE_LICENSE "BSD")
 set(CPACK_RPM_PACKAGE_URL "${SITE_URL}")
-if(AF_WITH_GRAPHICS)
+if(AF_BUILD_FORGE)
```
These are optional now. You don't need these libs to run ArrayFire but they should be marked optional. use CPACK_RPM_PACKAGE_SUGGESTS instead of CPACK_RPM_PACKAGE_REQUIRES
